### ISSUE TITLE
Iterate over enabled_mod_list.

### ISSUE THREAD

**Issue by ns-rse** [Mesnage-Org/pgfinder#113]
In `pgfinder/matching.py` the `data_analysis()` function has the following for creating a series of data frames for each modification and then concatenating them to `master_frame`....

```python
    if "Sodium" in enabled_mod_list:
        adducts_sodium_df = modification_generator(obs_theo_df, "Sodium")
    else:
        adducts_sodium_df = pd.DataFrame()

    if "Potassium" in enabled_mod_list:
        adducts_potassium_df = modification_generator(obs_theo_df, "Potassium")
    else:
        adducts_potassium_df = pd.DataFrame()

    if "Anh" in enabled_mod_list:
        anhydro_df = modification_generator(obs_theo_df, "Anh")
    else:
        anhydro_df = pd.DataFrame()

    if "DeAc" in enabled_mod_list:
        deacetyl_df = modification_generator(obs_theo_df, "DeAc")
    else:
        deacetyl_df = pd.DataFrame()

    if "DeAc_Anh" in enabled_mod_list:
        deac_anhy_df = modification_generator(obs_theo_df, "DeAc_Anh")
    else:
        deac_anhy_df = pd.DataFrame()
    if "O-Acetylated" in enabled_mod_list:
        oacetyl_df = modification_generator(obs_theo_df, "O-Acetylated")
    else:
        oacetyl_df = pd.DataFrame()

    if "Nude" in enabled_mod_list:
        nude_df = modification_generator(obs_theo_df, "Nude")
    else:
        nude_df = pd.DataFrame()

    if "Decay" in enabled_mod_list:
        decay_df = modification_generator(obs_theo_df, "Decay")
    else:
        decay_df = pd.DataFrame()

    if "Amidation" in enabled_mod_list:
        ami_df = modification_generator(obs_theo_df, "Amidated")
    else:
        ami_df = pd.DataFrame()

    if "Amidase" in enabled_mod_list:
        deglyco_df = modification_generator(obs_theo_df, "Amidase Product")
    else:
        deglyco_df = pd.DataFrame()

    if "Double_Anh" in enabled_mod_list:
        double_Anhydro_df = modification_generator(obs_theo_df, "Double_Anh")

    else:
        double_Anhydro_df = pd.DataFrame()

    master_list = [
        obs_theo_df,
        adducts_potassium_df,
        adducts_sodium_df,
        anhydro_df,
        deac_anhy_df,
        deacetyl_df,
        oacetyl_df,
        decay_df,
        nude_df,
        ami_df,
        deglyco_df,
        double_Anhydro_df,
    ]
    master_frame = pd.concat(master_list)
```

Rather than test for the presence of a modification and create empty pd.DataFrame() if none is found  we can put the original `obs_theo_df` into `master_list` and append all enabled modifications by iterating over them and _then_ `pd.concat()` to a list, greatly reducing the lines of code from multiple `if: ... else: ...` to the following...

```python
    master_list = [obs_theo_df]
    for modification in enabled_mod_list:
        master_list.append(modification_generator(obs_theo_df, modification))
    master_frame = pd.concat(master_list)
```

**Comment by ns-rse** [comment]
Hi @Ankur-AVP-Patel , @smesnage and @bobturneruk 

Before committing I ran the tests and the regression test fails.

This is because if the `enabled_mod_list` includes `Amidation` (which `test_regression.py` does) then `modification_generator()` is called with `Amidated` since we have the following in what I am trying to replace...

```python
416    if "Amidation" in enabled_mod_list:
417        ami_df = modification_generator(obs_theo_df, "Amidated")
418    else:
419        ami_df = pd.DataFrame()
```

But as my solution to reduce code and simplify things involves iterating over `enabled_mod_list` it means `Amidation` rather than `Amidated` is now passed to `modication_generator()`, and the test fails because no such key exists in parameterised `mod_type` (since its key is `Amidated` and not `Amidation`).

This feels to me that nomenclature is perhaps a bit out of sync, but `Amidation` being in `enabled_mod_list` and `Amidated` being used in the call to `modification_generator()` has been in there for [2 years](https://github.com/Mesnage-Org/pgfinder/blame/6c2ddb2d93fe90dfc1ba485e29fef3ac1923f1d5/pgfinder/matching.py) (scroll down to lines 363-367). I can see one of two possible solutions...

* Use `Amidated` in the `enabled_mod_list` of the test and all subsequent usage.
* Modify the `mod_type` dictionary key to be `Amidation` (this involves changing the parameterised values in `config/parameters.yaml`.

Advice on which is preferable would be gratefully received.

**Comment by bobturneruk** [comment]
`Amidated` seems more consistent with the other modification names, but I think this is sufficiently science-y to require @Ankur-AVP-Patel or @smesnage to make the call.

**Comment by smesnage** [comment]
Damn, you opened a Pandora box!
Your comment made me realise that there is a lot of inconsistency in the way we describe modifications in the output of the script and/or the choices offered for the Jupyter notebook search. I think the modifications below would be welcome:

1) Sodium searches could be amended
   - it should be "Sodium (Na+)" in the Jupyter notebook instead of "Sodium"
   - output is ok as **Na+** gm-AEJAA; NO CHANGE REQUIRED
  
2) Potassium searches could be amended
   - it should be "Potassium (K+)" in the Jupyter notebook instead of "Potassium"
   - output is ok as **K+** gm-AEJAA; NO CHANGE REQUIRED
  
3) Anhydro searches could be amended    
    - it should be "Anhydro (Anh)" in the Jupyter notebook instead of "Anh";
    - output should be gm-AEJAA **(Anh)**;  NO CHANGE REQUIRED

4) de-acetylation searches could be amended    
    - it should be "de-Acetylation (-Ac)" in the Jupyter notebook instead of "DeAc"; 
    - output should be gm-AEJAA **(-Ac)** instead of gm-AEJAA (DeAc);

5) Combined de-acetylation/anhydro searches could be amended    
    - it should be "de-Acetylation and Anhydro (-Ac, Anh)" in the Jupyter notebook instead of "DeAc_Anh"; 
    - output should be gm-AEJAA **(-Ac, Anh)** instead of gm-AEJAA (DeAc_Anh);

6) Nude searches could be amended    
    - it should be "Extra gm (gm-)" in the Jupyter notebook instead of "Nude"; 
    - output should be gm-gm-AEJAA;   NO CHANGE REQUIRED

7) Decay searches could be amended    
    - it should be "In-source decay" in the Jupyter notebook instead of "Decay"; 
    - output should be m-AEJAA;   NO CHANGE REQUIRED
   
8) Amidation searches could be amended    
    - it should be "Amidation" in the Jupyter notebook instead of "Amidated"; 
    - output should be gm-AEJAA (Amidated);   NO CHANGE REQUIRED

9) Amidase searches could be amended    
    - it should be "Loss of gm (Amidase product)" in the Jupyter notebook instead of "Amidase"; 
    - output should be AEJAA (Amidase product) instead of **gm-**AEJAA (Amidase product)

10) Double Anhydro searches could be amended    
    - it should be "Double Anhydro (Double Anh)" in the Jupyter notebook instead of "Double_Anh"; 
    - output should be gm-AEJAA (**Double Anh**) instead of gm-AEJAA (Double_Anh)

11) Multimers, multimers_Glyco and Multimers_Lac are ok EXCEPT 2 things:
    - change "multimers_Glyco" for "Multimers_Glyco " (add a capital);
    - output should be **gm**-AEJAA instead of **GM**-AEJAA (we changed the syntax to avoid confusion between g=GlcNAc and G=Glycine and m=MurNAc and M=methionine  

12) O-acetylation searches could be amended
    - it should be "O-Acetylation (+Ac)" in the Jupyter notebook instead of "O-Acetylated";
    - output should be gm-AEJAA (+Ac)  

Does this help? happy to answer specific questions if I missed the point.

Cheers

**Comment by ns-rse** [comment]
Thanks @smesnage nothing like opening up a can of :worm:!

As this issue is very specific I think that from 

> Amidation searches could be amended
>
> * it should be "Amidation" in the Jupyter notebook instead of "Amidated";
> * output should be gm-AEJAA (Amidated); NO CHANGE REQUIRED

I should change `Amidated` > `Amidation` in the `config/parameters/yaml`.

For the rest of the changes I think it is more sensible to create a new issue.

**Comment by bobturneruk** [comment]
@ns-rse - yes the specifics needed to fix this are now here, along with some more changes which should be split off into a sperate issue so they can be worked on seperately.

**Comment by ns-rse** [comment]
Afternoon, another question with this particular issue has arisen.

After converting `Amidated` > `Amidation` in `config/parameters.yaml` I hit another similar issue with `Amidase` not being found in `config/parameters.yaml` looking at this it appears to be because it is listed as `Amidase Product` and so I used my initiative and advice that what the user sees is what should be parameterised to change this to...

```
  Amidation:
    mass: -0.9840
    name: Amidated
  Amidase:
    mass: -480.1955
    name: (Amidase Product)
```

However, the `tests/test_regression.py` includes `multimers_Glyco` in the set of modifications to test (see lines `18-33` of [test_regression.py](https://github.com/Mesnage-Org/pgfinder/blob/master/tests/test_regression.py)) and unfortunately there is no good candidate for this in the `config/parameters.yaml` (nor the original code see lines 166-195 of that are removed from `matching.py` in [97](https://github.com/Mesnage-Org/pgfinder/pull/97/files#diff-c695423fdc281c09ad705c6b3bae990c8f3c15ec70f6cae6a9b08410d4dfc3e7)).

This means that the modification `multimers_Glyco` was never actually being tested because the logic used in `if:...else:` statements to generate dataframes for modifications and since there was no `if mod_type == "multimers_Glyco:` this was never caught.

Two possible resolutions, from my naive perspective...

1. There is a `mod_type` missing from `config_parameters.yaml` for the modification type `multimers_Glyco` in which case this can be added if we know the `mass` and `name` (if there is one.
2. The inclusion of `multimers_Glyco` in the `mod_list` of the regression test is erroneous and should be removed.

Or there is perhaps something else that I'm not aware of?

**Comment by smesnage** [comment]
All this is a bit beyond me so I'll cling on to the low hanging fruits; add brackets, so change
 
 Amidation:
    mass: -0.9840
    name: Amidated

for

 Amidation:
    mass: -0.9840
    name: (Amidated)

**Comment by ns-rse** [comment]
@smesnage duly added.

@Ankur-AVP-Patel has pointed out to me that `multimers_Glyco` is handled by lines prior to those I've replaced with...

```python
    elif "multimers_Glyco" in enabled_mod_list:
        LOGGER.info("Building multimers from obs muropeptides")
        theo_multimers_df = multimer_builder(obs_monomers_df, 1)
        LOGGER.info("Filtering theoretical multimers by observed")
        obs_multimers_df = filtered_theo(ff, theo_multimers_df, user_ppm)
```

...but this logic is different to how other items in the `enabled_mod_list` are handled as it uses first `multimer_builder()` to build the Theoretical Multimers DataDrame (`theo_multimers_df`) before then using `filtered_theo()` to build the Observed Multimers DataFrame (`obs_multimers_df`).

For this reason I think I will try spliting the testing for now into `monomers` and `multimers`.

**Comment by ns-rse** [comment]
The changes are relatively simple and have been made on branch [ns-rse/113-itterate-enabled_mod_list]() (see commit #5c85e14).

Because the `enabled_mod_list` contained both Monomers and Multimers that were to be processed and they required different logic/handling I have split this into `enabled_monomers` and `enabled_multimers` in the `data_analysis()` function (and updating the test to use these too).

Unfortunately the tests currently fail, there is one less line for some reason and I spent most of Friday trying to fathom it out but didn't get there. It may be simpler to work investigate/resolve with _after_ #104 has been completed and merged (which would require rebasing this branch once those changes have been merged).
